byway/src/main/webapp/recommendations.js

```diff
@@ -0,0 +1,197 @@
+    calcRoute(directionsService, directionsRenderer, start, end);
+  });
+  placesService = new google.maps.places.PlacesService(map);
+}
+
+/**
+ * Add start and end points of the route to regions to search around.
+ * Temporary set up to mimic adding several endpoints between legs.
+ */
+function addDestinations(start, end) {
+  destinations.add(start);
+  destinations.add(end);
+  regions.push(...destinations);
+}
+
+/**
+ * Creates a route between two points and loads onto the map.
+ * Finds points along the path to load the regions Array.
+ * Loads recommendations centered around these points.
+ * @param {DirectionsService directionsService} finds directions
+ * @param {DirectionsRenderer directionsRenderer} renders the route
+ * @param {LatLng start} starting point location
+ * @param {LatLng end} ending point location
+ */
+function calcRoute(directionsService, directionsRenderer, start, end) {
+  const request = {
+      origin:  start,
+      destination: end,
+      travelMode: 'DRIVING'
+  };
+  directionsService.route(request, function(response, status) {
+    if (status == 'OK') {
+      directionsRenderer.setDirections(response);
+      findRegions(response);
+      loadRecommendations();
+    } else {
+      alert("Could not calculate route due to: " + status);
+    }
+  });
+}
+
+/**
+ * Goes through steps along route to find the average location
+ * between a step's start and end location. Store these points
+ * in the regions Array.
+ * @param {DirectionsResult directionResult} contains directions
+ * for the route made.
+ */
+function findRegions(directionResult) {
+  const myRoute = directionResult.routes[0].legs[0];
+  for(step of myRoute.steps) {
+    const avgLat = (step.start_location.lat() + step.end_location.lat()) / 2;
+    const avgLng = (step.start_location.lng() + step.end_location.lng()) / 2;
+    const avgLoc = {lat: avgLat, lng: avgLng};
+    if(step.distance.value > MIN_DISTANCE) {
+      regions.push(avgLoc);
+    }
+  }
+}
+
+/**
+ * Set a timeout to delay the browser.
+ * @param {Number delayMs} number of milliseconds
+ */
+function delayPromise(delayMs) {
+  return new Promise(resolve => setTimeout(resolve, delayMs));
+}
+
+/**
+ * Go through a user's interests and search for places
+ * fitting those interests. Search around the regions
+ * previously found using textSearch.
+ */
+async function loadRecommendations() {
+  for(interest of interests) {
+    for(region of regions) {
+      await delayPromise(250);
+      const request = {
+        location: region,
+        radius: RADIUS,
+        query: interest
+      }
+      const results = await new Promise(resolve => {
+        placesService.textSearch(request, (result, status) => {
+          if(status == "OK") {
+            resolve(result);
+          } else if(status == "OVER_QUERY_LIMIT") {
+            alert("Showing limited results");
+          } else {
+            alert("Status: " + status);
```

Comment:
Note: if you never use `resolve` the promise will never complete, and the outer function will hang forever on the `await`. This means if you call this function more than once and have it hang, it will leak memory, as an increasing number of incomplete promises will be left sitting around waiting to finish, even though they never will. You can deal with this by taking not just `resolve` but also `reject`, so you can send an error to the `await`. Even if you don't handle that error, you at least cause the `await` block to throw, which aborts the function and doesn't leak memory.

```js
new Promise((resolve, reject) => {
// ...
if (status === "OK") {
  resolve(result);
} else {
  reject(new Error(status)); // or similar. What is "result" anyway when status != OK?
}
// ...
});

Reply:
I updated to catch the exception as well and alert the user of the message. The result is null when `status !== OK`, but it seems after awaiting, the newly-named `placesFound` (previously named `results`) is undefined rather than null. I also added a check to call addRecommendations only when it is defined to avoid issues.